### PR TITLE
Cross talk correction code for `build_evt()`

### PR DESCRIPTION
**Pull request by tdixon97**
This contains code to cross talk correct the energies inside the evt tier. The underlying operations are performed in the `evt/modules/cross_talk.py` module acting on `awkward` arrays. The correction is applied recursively in a simple way to GERDA.

- [x] basic implementation
- [x] unit tests
- [x] handling of logic of negative pulses (arbitary)
- [x] positive polarity cross talk
- [x] format energies to allow matrix solution
- [x] fast (matrix algebra) implementation
- [x] function to compute multiplicity

### CODE REVIEW

File: src/pygama/evt/modules/xtalk.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+from legendmeta.catalog import Props
+from lgdo import lh5, types
+from lgdo.lh5 import ls
+
+from pygama.hit.build_hit import _remove_uneeded_operations, _reorder_table_operations
+
+from .. import utils
+
+
+def build_tcm_id_array(
```
Comment (gipert) [inline src/pygama/evt/modules/xtalk.py:18]: The naming of the TCM stuff is confusing, please rename as:
- `tcm_id` if it's the rawids (`tcm.id`). Can also just use `rawid`
- `tcm_idx` if it's the row in the hit-oriented tier files (`tcm.idx`). Can also use `hit_idx` (this is the name in the evt files)
- `tcm_index` if it's the row of the TCM arrays themselves (this is used in the evt config to filter other fields)

File: src/pygama/evt/modules/geds.py [hunk 2]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    else:
+        raise ValueError(f"Unknown mode: {mode}")
+
+
+def apply_xtalk_correction_and_calibrate(
+    datainfo: utils.DataInfo,
+    tcm: utils.TCMData,
+    table_names: Sequence[str],
+    *,
+    mode: str,
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:127]: would rename to `return_mode`

File: src/pygama/evt/modules/geds.py [hunk 3]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    uncalibrated_energy_name: str,
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:128]:
```suggestion
    cal_energy_expr: str,
```

File: src/pygama/evt/modules/geds.py [hunk 4]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    calibrated_energy_name: str,
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:129]:
```suggestion
    cal_energy_expr: str,
```

File: src/pygama/evt/modules/geds.py [hunk 5]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    par_files: str | list[str],
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:130]:
```suggestion
    cal_pars_files: str | list[str],
```

File: src/pygama/evt/modules/geds.py [hunk 6]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    multiplicity_logic: str,
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:131]:
```suggestion
    multiplicity_expr: str,
```

File: src/pygama/evt/modules/geds.py [hunk 7]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    threshold: float = None,
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:132]:
```suggestion
    energy_thr: float = None,
```
thr or threshold

File: src/pygama/evt/modules/geds.py [hunk 8]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    xtalk_matrix_filename: str = "",
+    xtalk_rawid_name: str = "xtc/rawid_index",
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:134]:
```suggestion
    xtalk_rawid_obj: str = "xtc/rawid_index",
```

File: src/pygama/evt/modules/geds.py [hunk 9]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    xtalk_matrix_name: str = "xtc/xtalk_matrix_negative",
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:135]:
```suggestion
    xtalk_matrix_obj: str = "xtc/xtalk_matrix_negative",
```

File: src/pygama/evt/modules/geds.py [hunk 10]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    positive_xtalk_matrix_name: str = "xtc/xtalk_matrix_positive",
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:136]:
```suggestion
    positive_xtalk_matrix_obj: str = "xtc/xtalk_matrix_positive",
```

File: src/pygama/evt/modules/geds.py [hunk 11]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
...
+    out_param: str = None,
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:137]:
```suggestion
    recal_with_estimator: str = None,
```
?

File: src/pygama/evt/modules/xtalk.py (hunk 1 above)
Comment (gipert) [reply src/pygama/evt/modules/xtalk.py:18]: I would rename this function to sth like `gather_tcm_indices()` in analogy with `spms.gather_tcm_data()`?

File: src/pygama/evt/modules/xtalk.py [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+        )
+        tcm_ids = np.where(tcm.id == table_id)[0]
+        idx_events = ak.to_numpy(tcm.idx[tcm.id == table_id])
+        tcm_ids_out[idx_chan][idx_events] = tcm_ids
+
+    # transpose to return object where row is events and column rawid idx
+    return tcm_ids_out.T
+
+
+def build_energy_array(
```
Comment (gipert) [inline src/pygama/evt/modules/xtalk.py:48]:
```suggestion
def gather_energy(
```
maybe?

File: src/pygama/evt/modules/xtalk.py [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+        energies_out[idx_events, idx_chan] = res.nda
+
+    return energies_out
+
+
+def filter_hits(
+    datainfo: utils.DataInfo,
+    tcm: utils.TCMData,
+    logic: str,
+    corrected_energy: np.ndarray,
```
Comment (gipert) [inline src/pygama/evt/modules/xtalk.py:108]:
```suggestion
    filter_expr: str,
    xcorr_energy: np.ndarray,
```
?

File: src/pygama/evt/modules/xtalk.py [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+        mask[idx_events, idx_chan] = res.nda
+
+    return mask
+
+
+def xtalk_corrected_energy(
+    uncalibrated_energies: np.ndarray,
+    calibrated_energies: np.ndarray,
+    matrix: np.ndarray,
+    threshold: float = None,
```
Comment (gipert) [inline src/pygama/evt/modules/xtalk.py:175]:
```suggestion
def xtalk_correct_energy(
    uncal_energy: np.ndarray,
    cal_energy: np.ndarray,
    xtalk_matrix: np.ndarray,
    energy_thr: float = None,
```

File: src/pygama/evt/modules/xtalk.py [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+        energies_threshold = np.where(
+            calibrated_energies_no_nan < threshold, 0, uncalibrated_energies_no_nan
+        )
+    else:
+        energies_threshold = uncalibrated_energies_no_nan
+    energies_correction = -np.matmul(matrix, energies_threshold.T).T
+    return uncalibrated_energies + energies_correction
+
+
+def get_xtalk_correction(
```
Comment (gipert) [inline src/pygama/evt/modules/xtalk.py:213]: does this really need to be split from the funcion above?

File: src/pygama/hit/build_hit.py [hunk 16]
```diff
@@ -259,3 +259,37 @@ def _one_pass(config):
             return new
         else:
             current = new
+
+
+def _get_dependencies(config, par, pars=None):
```
Comment (gipert) [inline src/pygama/hit/build_hit.py:264]: add a short docstring to these two funcs

File: src/pygama/evt/modules/geds.py [hunk 17]
```diff
@@ -4,10 +4,14 @@
...
 from collections.abc import Sequence
 
+import awkward as ak
 import numpy as np
 from lgdo import lh5, types
 
 from .. import utils
+from . import xtalk
+
+sto = lh5.LH5Store()
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:14]: don't use a store unless you need to keep files open

File: src/pygama/evt/modules/geds.py [hunk 18]
```diff
@@ -41,33 +45,176 @@ def apply_xtalk_correction(
     tcm: utils.TCMData,
     table_names: Sequence[str],
     *,
-    energy_observable: types.VectorOfVectors,
-    rawids: types.VectorOfVectors,
-    xtalk_matrix_filename: str,
+    mode: str,
```
Comment (gipert) [inline src/pygama/evt/modules/geds.py:48]: see comments below about arg names

File: src/pygama/evt/modules/xtalk.py (hunk 1 above)
Comment (tdixon97) [reply src/pygama/evt/modules/xtalk.py]: changed it to ``tcm_index'' throughout

File: src/pygama/hit/build_hit.py (hunk 16 above)
Comment (tdixon97) [reply src/pygama/hit/build_hit.py:264]: @ggmarshall  ?

File: src/pygama/evt/modules/xtalk.py (hunk 15 above)
Comment (gipert) [reply src/pygama/evt/modules/xtalk.py:213]: ok let's maybe rename the above to `xtalk_correct_energy_impl` or similar?

File: src/pygama/hit/build_hit.py (hunk 16 above)
Comment (ggmarshall) [reply src/pygama/hit/build_hit.py:264]: sure I'll see what copilot comes up with

Comment (gipert) [reply src/pygama/hit/build_hit.py:264]: maybe `_this_gets_dependencies_i_suppose()`?

**Pull request merged**